Name: <redacted> <redacted>

GitHub handle's username: <redacted>

The URL of your repo (e.g., github.com/YOUR_SUBMISSION): https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>

The commit ID of your submission: 8628b7c

In [1]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
import os 
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_iris
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier

# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely or locally)
iris = load_iris()
X = iris.data
y = iris.target

iris_df = pd.DataFrame(data=X, columns=iris.feature_names)
iris_df['target'] = y

# Output the first 15 rows of the data
print(iris_df.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print("\nSummary of the dataset:")
print(iris_df.info())
print("\nDescriptive statistics:")
print(iris_df.describe())

    sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4 

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

As mentioned in the readme, the Iris dataset is a classic used in the field of ml/data science as an example a lot. I, myself, have encountered this very dataset before. 

The features (all numerical) of this data set are include: 
- Sepal length: length of the sepals (the outer part of the flower) measured in centimeters
- Sepal Width: width of the sepals measured in centimeters
- Petal Length: length of the petals (the inner part of the flower) measured in centimeters
- Petal Width: width of the petals measured in centimeters

The labels are the type of iris flower each sample belong to (denotes a class/type of species). They are:
- Iris Setosa: represented as integer value 0 in the dataset
- Iris Versicolor: represented as integer value 1 in the dataset
- Iris Virginica: represented as integer value 2 in the dataset

The numeric labels serve as targets for machine learning algorithms, allowing them to learn and make predictions based on the provided features. This dataset has been extensively used for testing and demonstrating classification algorithms because of its simplicity and clear class separation.


# Part 2: Split the dataset into train and test

In [3]:
# Take the dataset and split it into our features (X) and label (y)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
print("Training set shape:", X_train.shape)
print("Test set shape:", X_test.shape)
print("Training labels shape:", y_train.shape)
print("Test labels shape:", y_test.shape)

Training set shape: (135, 4)
Test set shape: (15, 4)
Training labels shape: (135,)
Test labels shape: (15,)


# Part 3: Logistic Regression

In [4]:
# i. Use sklearn to train a LogisticRegression model on the training set
log_reg = LogisticRegression(max_iter=200)
log_reg.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_datapoint = X_test[0].reshape(1, -1)  # Reshape to create a 2D array with one sample
predicted_probabilities = log_reg.predict_proba(sample_datapoint)

print("Predicted probabilities for the sample datapoint:", predicted_probabilities)

# iii. Report on the score for Logistic regression model, what does the score measure?
y_pred = log_reg.predict(X_test)
score = accuracy_score(y_test, y_pred)

print("Accuracy score of the Logistic Regression model:", score)

# iv. Extract the coefficents and intercepts for the boundary line(s)
coefficients = log_reg.coef_
intercept = log_reg.intercept_

print("Coefficients for the boundary lines:\n", coefficients)
print("Intercept for the boundary lines:\n", intercept)

Predicted probabilities for the sample datapoint: [[0.00348709 0.83712491 0.159388  ]]
Accuracy score of the Logistic Regression model: 1.0
Coefficients for the boundary lines:
 [[-0.42600627  0.97180378 -2.44299304 -1.02937946]
 [ 0.51251837 -0.22353981 -0.21626022 -0.85201672]
 [-0.0865121  -0.74826397  2.65925326  1.88139618]]
Intercept for the boundary lines:
 [  9.49385005   1.91823857 -11.41208863]


**Report on the score for Logistic regression model, what does the score measure?**

I chose to use the accuracy score as a measure for this model because of its simplicity and interpretability. 

The accuracy score is 1. This indicates that the model correctly classified all instances in the test set, reflecting its effectiveness in distinguishing between the three iris species based on the provided features. Given that the Iris dataset exhibits clear class separability and lacks significant class imbalance, accuracy serves as a reliable measure of overall model performance, showcasing how well the logistic regression model captures the underlying relationships in the data.

Other performance metrics/indicators like the predicted probabilities for a sample datapoint further illustrate the model's capability to assess class membership with confidence. With predicted probabilities of approximately 0.837 for Iris Versicolor, the model indicates strong confidence in this classification, suggesting an effective decision boundary has been learned. The coefficients and intercepts from the logistic regression provide insights into the influence of each feature on class probabilities, demonstrating how the model leverages the provided measurements to make informed predictions. This combination of high accuracy and informative outputs emphasizes the appropriateness of logistic regression for this task.

# Part 4: Support Vector Machine

In [5]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svm = SVC(probability=True)  # Set probability=True to enable probability estimates
svm.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_datapoint = X_test[0].reshape(1, -1)  # Reshape to create a 2D array with one sample
predicted_probabilities = svm.predict_proba(sample_datapoint)

print("Predicted probabilities for the sample datapoint:", predicted_probabilities)

# iii. Report on the score for the SVM, what does the score measure?
y_pred = svm.predict(X_test)
score = accuracy_score(y_test, y_pred)

print("Accuracy score of the Support Vector Classifier:", score)
# score measures the proportion of correctly classified samples


Predicted probabilities for the sample datapoint: [[0.00686141 0.89883721 0.09430137]]
Accuracy score of the Support Vector Classifier: 1.0


**Report on the score for the SVM, what does the score measure?**

The accuracy score of 1.0 achieved by the Support Vector Classifier (SVC) indicates perfect classification of the samples in the test set, demonstrating the model's effectiveness in distinguishing between the three species of iris flowers. An accuracy score measures the proportion of correct predictions made by the model compared to the total predictions, providing a straightforward assessment of overall performance. Given the clear separation of classes in the Iris dataset, this high accuracy reflects the SVC's capability to learn the optimal decision boundaries effectively, leveraging its powerful kernel methods that can manage complex relationships between features.

Additionally, the predicted probabilities for a sample datapoint, showing approximately `[0.0069, 0.8988, 0.0943]`, indicate that the SVC assigns a significantly high likelihood (about 89.88%) for the sample being classified as Iris Versicolor, suggesting high confidence in this prediction. The low probability for Iris Setosa highlights the model's ability to distinguish between the classes accurately. By outputting not just class predictions but also probabilities, the SVC provides insights into its confidence level in those predictions, which can be valuable in applications where understanding certainty is important. Overall, the SVC demonstrates excellent performance on this dataset, supporting its reputation as a robust and flexible classification algorithm.

# Part 5: Neural Network

In [6]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp = MLPClassifier(hidden_layer_sizes=(10, 10), max_iter=1000, random_state=42)
mlp.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_datapoint = X_test[0].reshape(1, -1)  # Reshape to create a 2D array with one sample
predicted_probabilities = mlp.predict_proba(sample_datapoint)

print("Predicted probabilities for the sample datapoint:", predicted_probabilities)

# iii. Report on the score for the Neural Network, what does the score measure?
y_pred = mlp.predict(X_test)
score = accuracy_score(y_test, y_pred)

print("Accuracy score of the Neural Network (MLP Classifier):", score)
# score measures the proportion of correctly classified samples

# iv: Experiment with different options for the neural network, report on your best configuration

# Define different configurations to test
configurations = [
    {'hidden_layer_sizes': (5,), 'max_iter': 500},
    {'hidden_layer_sizes': (10, 5), 'max_iter': 500},
    {'hidden_layer_sizes': (20, 10), 'max_iter': 1000},
    {'hidden_layer_sizes': (10, 10), 'max_iter': 1000, 'activation': 'tanh'},
    {'hidden_layer_sizes': (10, 10), 'max_iter': 1000, 'activation': 'relu'},
]

best_score = 0
best_config = None

# Iterate through each configuration
for config in configurations:
    print(f"Testing configuration: {config}")
    mlp = MLPClassifier(**config, random_state=42)
    mlp.fit(X_train, y_train)
    y_pred = mlp.predict(X_test)
    score = accuracy_score(y_test, y_pred)
    print("Accuracy score:", score)
    
    # Keeping track of the best configuration
    if score > best_score:
        best_score = score
        best_config = config

print("\nBest configuration:", best_config)
print("Best accuracy score:", best_score)


Predicted probabilities for the sample datapoint: [[0.00356211 0.95603497 0.04040291]]
Accuracy score of the Neural Network (MLP Classifier): 0.9333333333333333
Testing configuration: {'hidden_layer_sizes': (5,), 'max_iter': 500}


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


Accuracy score: 0.7333333333333333
Testing configuration: {'hidden_layer_sizes': (10, 5), 'max_iter': 500}


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


Accuracy score: 0.9333333333333333
Testing configuration: {'hidden_layer_sizes': (20, 10), 'max_iter': 1000}
Accuracy score: 1.0
Testing configuration: {'hidden_layer_sizes': (10, 10), 'max_iter': 1000, 'activation': 'tanh'}
Accuracy score: 0.9333333333333333
Testing configuration: {'hidden_layer_sizes': (10, 10), 'max_iter': 1000, 'activation': 'relu'}
Accuracy score: 0.9333333333333333

Best configuration: {'hidden_layer_sizes': (20, 10), 'max_iter': 1000}
Best accuracy score: 1.0


**Report on the score for the Neural Network, what does the score measure?**

The accuracy score of approximately 0.933 for the Neural Network (MLP Classifier) indicates that the model successfully classified about 93.3% of the samples in the test set correctly. This metric measures the proportion of correct predictions made by the model relative to the total number of predictions, serving as a straightforward evaluation of the overall effectiveness of the model. While this score reflects strong performance, it suggests that there were some misclassifications, highlighting areas where the model may improve its learning, particularly given the linear separability of the Iris dataset, where we might expect high accuracy across various models.

The predicted probabilities for the sample datapoint, which are approximately `[0.0036, 0.9560, 0.0404]`, illustrate that the neural network assigns a high likelihood of around 95.6% to the sample being classified as Iris Versicolor. This high probability reflects the model's confidence in its prediction, indicating that it effectively learned the underlying patterns within the dataset. The low probabilities for the other classes show that the model can differentiate between species effectively. Thus, while the accuracy score indicates room for improvement, the neural network's ability to produce confident predictions enhances its utility for applications needing probabilistic insights.

**Experiment with different options for the neural network, report on your best configuration**

In the experimentation with different configurations of the neural network (MLP Classifier), several setups were tested to optimize model performance. The initial configuration with a single hidden layer of 5 neurons for 500 iterations resulted in an accuracy score of approximately 0.733, indicating that the model had not effectively learned the patterns in the data. A second configuration, which included two hidden layers consisting of 10 and 5 neurons, yielded a significantly improved accuracy score of 0.933, showing better data representation. However, a convergence warning signaled that the model had not fully optimized within the specified iterations, prompting further experimentation with additional configurations.

Ultimately, the best-performing configuration was identified as having two hidden layers with 20 and 10 neurons, trained for 1000 iterations, which achieved a perfect accuracy score of 1.0. This suggests that the deeper architecture allowed the model to capture more complex relationships in the data, leading to optimal performance. While other configurations provided solid accuracy rates of approximately 0.933, the winning configuration demonstrated superior learning capability, highlighting the importance of network design and iteration count in training effective neural networks. This exercise underscores the neural network's flexibility and adaptability, allowing for fine-tuning to achieve the best results for the task at hand.

# Part 6: K-Nearest Neighbors

In [7]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

knn = KNeighborsClassifier(n_neighbors=3)  # You can adjust n_neighbors as needed
knn.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

sample_datapoint = X_test[0].reshape(1, -1)  # Reshape to create a 2D array with one sample
predicted_probabilities = knn.predict_proba(sample_datapoint)

print("Predicted probabilities for the sample datapoint:", predicted_probabilities)

# iii. Report on the score for kNN, what does the score measure?

y_pred = knn.predict(X_test)
score = accuracy_score(y_test, y_pred)

print("Accuracy score of the k-Neighbors Classifier:", score)
# ccore measures the proportion of correctly classified samples

Predicted probabilities for the sample datapoint: [[0. 1. 0.]]
Accuracy score of the k-Neighbors Classifier: 1.0


**Report on the score for kNN, what does the score measure?**

The accuracy score of 1.0 for the k-Neighbors Classifier (kNN) indicates that the model correctly classified all instances in the test set, achieving perfect classification. This score measures the proportion of correct predictions made by the model out of the total predictions, providing a clear and direct assessment of the classifier's overall effectiveness. The perfect accuracy suggests that the kNN algorithm was successful in leveraging the proximity of data points in feature space to assign the correct class labels, reflecting the well-defined boundaries between the species in the Iris dataset.

Additionally, the predicted probabilities for the sample datapoint, shown as `[[0. 1. 0.]]`, indicate that the model assigns a 100% probability to the sample belonging to the second class (likely Iris Versicolor), with no overlap with the other classes. This certainty further confirms the model's ability to effectively distinguish between the flower species. The kNN classifier's non-parametric nature, which enables it to adapt to the local structure of the data, especially shines in this context, highlighting its effectiveness in scenarios where class boundaries are clear and well-separated, such as in the Iris dataset. Overall, the high accuracy and confident probability predictions underscore kNN's capabilities as a robust classification method.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

Throughout the notebook, we trained and evaluated several machine learning models on the Iris dataset, which consists of flower measurements used for classification into three species: Iris Setosa, Iris Versicolor, and Iris Virginica. The models we examined include:
- Logistic Regression
- Support Vector Classifier (SVC)
- Neural Network (MLP Classifier)
- k-Neighbors Classifier (KNN)

**Model Performance**

Logistic Regression: This model provided high accuracy, indicating that it distinguished well among the classes. Its linear nature works effectively due to the relatively straightforward decision boundaries in the dataset.

Support Vector Classifier (SVC): SVC also achieved a perfect classification score. It typically performs well in high-dimensional spaces by focusing on support vectors which define the decision boundary, making it likely to succeed with the Iris dataset.

Neural Network (MLP Classifier): The neural network demonstrated strong performance and flexibility, allowing for multiple configurations. Several attempts provided perfect accuracy, showing that it could easily model the relationships in the data.

k-Neighbors Classifier (KNN): KNN also yielded high accuracy, particularly when choosing an optimal number of neighbors (n_neighbors). Given that KNN is a non-parametric classifier, it adapts to local data structures effectively, which is beneficial for the simplified nature of the Iris dataset.

**Best Performing Model(s)**

While most models performed equally well in terms of accuracy (often 100%), the best-performing models were likely the Support Vector Classifier (SVC) and Neural Network (MLP Classifier) due to their ability to handle complex boundaries and flexibility in accommodating different architecture configurations. However, MLP classifier was a little lower in terms of accuracy vs the SVM model (0.933 vs 1).

**Reasons for Performance**

The reason these models performed well is likely due to the following:
- Simplicity of the Dataset: The Iris dataset is relatively small and simple, allowing most classification algorithms to achieve high accuracy without overfitting. The distinct classes are well-separated, making it easier for models to learn and generalize.
- Training Data: With 90% of the data used for training, the models had sufficient information to learn effective classification boundaries.
- Model Adaptability: SVC and MLP Classifier can capture more complex relationships in the data due to their nonlinear capabilities, whereas logistic regression is limited to linear relationships (albeit still effective here).

**Surprises in the Exercise**
High Accuracy Across Models: It was surprising to see that several different models achieved perfect accuracy on a relatively simple dataset. This highlights the effectiveness of various algorithms, even those based on very different principles (e.g., linear vs. non-linear classifiers).

KNN's Non-Training Nature: Understanding that KNN is non-parametric and simply stores all training data for reference was enlightening. This contrasts starkly with models that require explicit learning and parameter estimation.

Flexibility of Neural Networks: The ability to easily modify and experiment with different configurations in the neural network demonstrated how adaptable machine learning can be and how many hyperparameters can influence performance.

**Conclusion**

Overall, this exercise provided valuable insights into different classification algorithms, their strengths and weaknesses, and how they perform on a straightforward dataset like Iris. The uniform high performance underscores the importance of model selection and adaptability depending on the dataset characteristics. However, it also serves as a reminder to evaluate the performance on more complex datasets to assess model robustness and generalization capabilities effectively.